airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/fileformatparser.py

```diff
@@ -132,49 +139,116 @@ def is_binary(self):
+
         :param json_schema: if this is passed in, pyarrow will attempt to enforce this schema on read, defaults to None
         """
         check_utf8 = True if self._format.get("encoding", "utf8").lower().replace("-", "") == "utf8" else False
         convert_schema = self.json_schema_to_pyarrow_schema(json_schema) if json_schema is not None else None
-        return pa.csv.ConvertOptions(
-            check_utf8=check_utf8, column_types=convert_schema, **json.loads(self._format.get("additional_reader_options", "{}"))
-        )
+        return {
+            **{"check_utf8": check_utf8, "column_types": convert_schema},
+            **json.loads(self._format.get("additional_reader_options", "{}")),
+        }
 
     def get_inferred_schema(self, file: Union[TextIO, BinaryIO]) -> dict:
         """
         https://arrow.apache.org/docs/python/generated/pyarrow.csv.open_csv.html
-        Note: this reads just the first block (as defined in _read_options() block_size) to infer the schema
+
+        This now uses multiprocessing in order to timeout the schema inference as it can hang.
+        Since the hanging code is resistant to signal interrupts, threading/futures doesn't help so needed to multiprocess.
+        https://issues.apache.org/jira/browse/ARROW-11853?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel
         """
-        streaming_reader = pa_csv.open_csv(file, self._read_options(), self._parse_options(), self._convert_options())
-        schema_dict = {field.name: field.type for field in streaming_reader.schema}
+
+        def get_schema(file_sample: str, read_opts: dict, parse_opts: dict, convert_opts: dict) -> dict:
+            """
+            we need to reimport here to be functional on Windows systems since it doesn't have fork()
+            https://docs.python.org/3.7/library/multiprocessing.html#contexts-and-start-methods
+            """
+            import tempfile
+
+            import pyarrow as pa
+
+            # writing our file_sample to a temporary file to then read in and schema infer as before
+            with tempfile.TemporaryFile() as fp:
+                fp.write(file_sample)
+                fp.seek(0)
+                streaming_reader = pa.csv.open_csv(
+                    fp, pa.csv.ReadOptions(**read_opts), pa.csv.ParseOptions(**parse_opts), pa.csv.ConvertOptions(**convert_opts)
+                )
+                schema_dict = {field.name: field.type for field in streaming_reader.schema}
+            return schema_dict
+
+        # boto3 stuff can't be pickled and so we can't multiprocess with the actual fileobject on Windows systems
+        # we're reading block_size*2 bytes here, which we can then pass in and infer schema from block_size bytes
+        # the *2 is to give us a buffer as pyarrow figures out where lines actually end so it gets schema correct
+        file_sample = file.read(self._read_options()["block_size"] * 2)
+        schema_dict = None
```

Comment:
for readability could we refactor this bit to be: 

```
file_sample = file.read(self._read_options()["block_size"] * 2)
        schema_dict = self.run_in_external_process(get_schema, file_sample, self._read_options(), self._parse_options(), self._convert_options())

        return self.json_schema_to_pyarrow_schema(schema_dict, reverse=True)
```

basically taking all the external process logic outside into a helper method? 

I tried it in my iDE, here's what I had for reference in case it helps: 

```
def run_in_external_process(self, fn, *args):
        schema_dict = None
        fail_count = 0
        while schema_dict is None:
            q_worker = mp.Queue()
            proc = mp.Process(
                target=multiprocess_queuer,
                args=(
                    dill.dumps(fn),  # use dill to pickle the get_schema function for Windows-compatibility
                    q_worker,
                    *args,
                ),
            )
            proc.start()
            try:
                # this attempts to get return value from function with a 20 second timeout
                return q_worker.get(timeout=20)
            except mp.queues.Empty:
                fail_count += 1
                if fail_count > 2:
                    raise TimeoutError("Timed out 3 times while trying to infer schema")
                self.logger.info("timed out on schema inference, retrying...")
            finally:
                try:
                    proc.terminate()
                except Exception as e:
                    self.logger.info(f"infer schema proc unterminated, error: {e}")
```

Reply:
yeah like it, changed!